### PR TITLE
fast-track: Add multipart EntityPart tests

### LINKED ISSUE #962: TCK test for multipart/form-data API
Related to issue #418 and PR #948

### PR DESCRIPTION
**Pull request by jim-krueger**
Fixes #962

### CODE REVIEW

**Review by jamezp — APPROVED** [review]
Looks good to me.

**Review by spericas — COMMENTED** [review]
I haven't looked at the styling in the recently added TCKs in detail, but the general indentation and styling in this PR seems very different to the rest of the codebase. Any chance we can make this more consistent?

**Review by jamezp — DISMISSED** [review]

**Comment by spericas** [comment]
> @spericas I changed tabs to spaces. Was there other style or formatting issues? Thanks

Thanks. In general, I think we would prefer this:

```
                    EntityPart.withName("file")
                            .content("test file", xmlFile())
                            .mediaType(MediaType.APPLICATION_XML)
                            .build());
```
 over this,
```
                    EntityPart.withName("file")
                    .content("test file", xmlFile())
                    .mediaType(MediaType.APPLICATION_XML)
                    .build()
                    );
```
in fluent calls.

However, I don't know how consistent we are on this. Maybe we can clean this up and set up some better checkstyle rules after the 4.0 release.

**Comment by jim-krueger** [comment]
@spericas Agreed, I will make those changes.   Also, I'm working through an issue related to the handling of Headers that will likely result in some additional changes.

**Review by jamezp — APPROVED** [review]
Looks good to me and the test passes on RESTEasy with WildFly.

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jansupol) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:190]: I think `test//single-form-param` should be `test/single-form-param`.

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jamezp) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:363]: I think it is required by spec on section [4.2.4](https://jakarta.ee/specifications/restful-ws/3.1/jakarta-restful-ws-spec-3.1#standard_entity_providers)

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jamezp) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:190]: +1 to this, nice catch!

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jansupol) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:243]: response.getStatusInfo() is a `StatusType`, which is not equal to `Status`

**Review by jansupol — CHANGES_REQUESTED** [review]

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jansupol) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:363]: @jamezp Can you please be more specific? This is `application/xml`.

Comment (jamezp) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:363]: @jansupol Sure:
> **4.2.4. Standard Entity Providers**
> An implementation MUST include pre-packaged MessageBodyReader and MessageBodyWriter implementations for the following Java and media type combinations:

In the list it has:
> `java.util.List<EntityPart>`
>  Multipart data (`multipart/form-data`)

I thought that meant you need a `MessageBodyReader` and a `MessageBodyWriter` that handled the type `List<EntityPart>`.

Comment (jansupol) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:363]: Still, `EntityPart.content(EntityPart)` is not supported, is it? The Spec does not allow to send a single `EntityPart`, only a list of it.

Comment (jansupol) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:363]: But with a wrong media type, media-type should be `MediaType.MULTIPART_FORM_DATA`

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jansupol) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:139]: Again, `StatusType` is not equal to `Status`

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jamezp) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:363]: Sorry, I understand now. It's the effective `EntityPart.withName("received-file").content(EntityPart)` that is the issue. That makes more sense, sorry for the confusion :)

It works in RESTEasy, but that doesn't mean any other implementation will work like that. I can understand the requirement for this change now.

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:190]: thanks

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:139]: Agreed

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:243]: agreed

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:363]: @jansupol  Please check it out again.   thanks

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jansupol) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:152]: The javadoc of getContent() says:
> Subsequent invocations will result in an IllegalStateException. Likewise this method will throw an IllegalStateException if it is called after calling getContent or getContent(GenericType).

But the getContent() has already been called by getMessage(). It looks like RestEasy allows for reading getContent() multiple times. Does it make sense to add a negative test to check the getContent has not been invoked multiple times?

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jansupol) [inline jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:381]: The same here: `test-string` EntityPart has content read three times: once for string, once for EntityPart and once for the InputStream. But the javadoc mandates to throw IllegalStateException.

Comment (jansupol) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:381]: I wonder how that's implemented in RestEasy, because there would need to be some automatic buffering of the InputStream, but that's gonna bite you for large data.

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (spericas) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:152]: Good catch

Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:152]: Yes, good catch.  I will fix the test and your suggestion to add a negative test is appropriate.

Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:152]: @jansupol Looking at the  [getContent javadoc](https://jakarta.ee/specifications/restful-ws/3.1/apidocs/jakarta.ws.rs/jakarta/ws/rs/core/entitypart#getContent()), shouldn't the no-parameter method also throw IllegalStateException since it could be called twice or after one of the other overloads?

Comment (jansupol) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:152]: I'd say not necessarily, since it returns the InputStream (for instance the ServletInputStream on the server side), which is meant to be the native datatype of the wired data, so the InputStream can be returned multiple times, although it can be already consumed, so the stream is not expected to be consumed twice, unlike with converting it to other data types (such as String) which are not supposed to be empty.

Comment (jansupol) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:152]: But sure, there is a discrepancy to getContent(InputStream.class) which throws, so I would not mind the no-arg method to throw, too. Or the InputStream can be excluded from throwing the ISE. Either way is possible.

**Comment by mkarg** [comment]
To allow all vendors to run the new test without sacrificing their week schedule, it might be good to remove the "fast-track" marker. In the end, a missing test is not a bug, and I voted -1 for fast-tracking non-bugs.

**Comment by jim-krueger** [comment]
From the original Committer Conventions:  
`(2) For non-API, non-spec, non-javadoc changes (e.g., pom, travis, checkstyle, etc) a proposal to fast track the PR is requested at submission time. If a fast tracked PR receives 3 committer +1 votes (and no -1), it can be merged immediately provided at least 1 day has passed since its submission.`

This is a TCK change and therefore "non-API, non-spec, and non-javadoc" so it can be fast-tracked.  All venders would only be required to run with this after it has been placed in a release and we are weeks (or more) away from that.   That being said, this PR is generating enough discussion that it is unlikely to be merged in a fast-tracked fashion.

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:153]: The changes I've made this afternoon resolve this issue and pass on RESTEasy after @jamezp made the appropriate changes.

File: jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java
Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:402]: I believe this is no longer an issue

Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:402]: How about taking the "singleParamTest" resource method and splitting it into 3, each with one `@FormParam` (one EntityPart, one String, and one InputStream) and then have the `singleFormParamTest()` invoke each separately, using the same `List<EntityPart>`?

Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:402]: Or maybe this test isn't all that useful or needed?

Comment (jansupol) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:402]: Splitting the test into three would be basically the same as the `multiFormParamTest`, would it not?
The important part is whether the spec was meant to support multiple `@FormParam("test-string")`. My understanding is it was not, because the content stream needs to be read (and buffered) multiple times, but perhaps it was meant so for the resources.

Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:402]: Yes, not much difference at all.    I'm leaning toward removal of that test and potentially clarifying the spec in this respect for the next release if we choose to go that route.

Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java:402]: That being said, can we agree that there also is currently not a spec (or javadoc) requirement for an IllegalStateException if user attempts to do something like what is in this test.   It is nebulous... they might get an exception or they might not depending on the the implemenation.

Comment (jim-krueger) [reply jaxrs-tck/src/main/java/ee/jakarta/tck/ws/rs/jaxrs31/ee/multipart/MultipartSupportIT.java]: I've removed the test.   We can revisit and potentially augment the spec for V 5.0.

**Comment by jim-krueger** [comment]
@jamezp @spericas @jansupol @mkarg Are there any more thoughts/concerns about the contents of this PR?   Thanks

**Comment by spericas** [comment]
In general it LGTM but I'll defer to @jansupol on this one

**Review by jamezp — APPROVED** [review]
LGTM.

**Review by spericas — APPROVED** [review]

**Pull request merged**
